plutus-core/plutus-core/src/PlutusCore/Check/Scoping.hs

```diff
@@ -353,29 +375,61 @@ so we only need to consider three more relations:
 
 The last two ensure that no new name has an old name's unique.
 -}
-checkScopeInfo :: ScopeInfo -> Either ScopeError ()
-checkScopeInfo scopeInfo = do
+checkScopeInfo :: BindingRemoval -> ScopeInfo -> Either ScopeError ()
+checkScopeInfo bindRem scopeInfo = do
     let disappearedBindings       = to DisappearedBindings       scopeInfo
         disappearedVariables      = to DisappearedVariables      scopeInfo
         appearedBindings          = to AppearedBindings          scopeInfo
         appearedVariables         = to AppearedVariables         scopeInfo
         stayedOutOfScopeVariables = to StayedOutOfScopeVariables scopeInfo
         stayedFreeVariables       = to StayedFreeVariables       scopeInfo
-    checkEmpty OldBindingsDiscordWithOutOfScopeVariables $
-        disappearedBindings `symmetricDifference` stayedOutOfScopeVariables
-    checkEmpty OldBindingsDiscordWithBoundVariables $
-        disappearedBindings `symmetricDifference` disappearedVariables
-    checkEmpty NewBindingsDiscordWithBoundVariables $
-        appearedBindings `symmetricDifference` appearedVariables
-    checkEmpty OldBindingsClashWithFreeVariables $
-        disappearedBindings `Set.intersection` stayedFreeVariables
-    checkEmpty OldBindingsClashWithNewBindings $
-        appearedBindings  `Set.intersection` disappearedBindings
-    checkEmpty NewBindingsClashWithFreeVariabes $
-        appearedBindings `Set.intersection` stayedFreeVariables
+    unless (bindRem == BindingRemovalOk) $ do
```

Comment:
> Use a `Writer` of problems rather than `Either`. This makes sense since finding a problem doesn't stop you finding more problems!

Yeah, I put a similar comment into the file when I originally wrote it:

```
-- We might want to use @Validation@ or something instead of 'Either'.
```

I agree with you, just don't feel like refactoring the machinery...